connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala

```diff
@@ -0,0 +1,153 @@
+ * before running the test. An example workflow with SBT for this test:
+ *   1. Compatibility test has reported an unexpected client API change.
+ *   1. Fix the wrong client API.
+ *   1. Build the client jar: `sbt package`
+ *   1. Run the test again: `sbt "testOnly
+ *      org.apache.spark.sql.connect.client.CompatibilitySuite"`
+ */
+class CompatibilitySuite extends AnyFunSuite { // scalastyle:ignore funsuite
+
+  private lazy val clientJar: File =
+    findJar(
+      "connector/connect/client/jvm",
+      "spark-connect-client-jvm-assembly",
+      "spark-connect-client-jvm")
+
+  private lazy val sqlJar: File = findJar("sql/core", "spark-sql", "spark-sql")
+
+  /**
+   * MiMa takes an old jar (sql jar) and a new jar (client jar) as inputs and then reports all
+   * incompatibilities found in the new jar. The incompatibility result is then filtered using
+   * include and exclude rules. Include rules are first applied to find all client classes that
+   * need to be checked. Then exclude rules are applied to filter out all unsupported methods in
+   * the client classes.
+   */
+  test("compatibility MiMa tests") {
+    val mima = new MiMaLib(Seq(clientJar, sqlJar))
+    val allProblems = mima.collectProblems(sqlJar, clientJar, List.empty)
+    val includedRules = Seq(
+      IncludeByName("org.apache.spark.sql.Column"),
+      IncludeByName("org.apache.spark.sql.Column$"),
+      IncludeByName("org.apache.spark.sql.Dataset"),
+      // TODO(SPARK-42175) Add the Dataset object definition
+      // IncludeByName("org.apache.spark.sql.Dataset$"),
+      IncludeByName("org.apache.spark.sql.DataFrame"),
+      IncludeByName("org.apache.spark.sql.SparkSession"),
+      IncludeByName("org.apache.spark.sql.SparkSession$")) ++ includeImplementedMethods(clientJar)
+    val excludeRules = Seq(
+      // Filter unsupported rules:
+      // Two sql overloading methods are marked experimental in the API and skipped in the client.
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.SparkSession.sql"),
+      // Skip all shaded dependencies in the client.
+      ProblemFilters.exclude[Problem]("org.sparkproject.*"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.connect.proto.*"))
+    val problems = allProblems
+      .filter { p =>
+        includedRules.exists(rule => rule(p))
+      }
+      .filter { p =>
+        excludeRules.forall(rule => rule(p))
+      }
+
+    if (problems.nonEmpty) {
+      fail(
+        s"\nComparing client jar: $clientJar\nand sql jar: $sqlJar\n" +
+          problems.map(p => p.description("client")).mkString("\n"))
+    }
+  }
+
+  test("compatibility API tests: Dataset") {
+    val clientClassLoader: URLClassLoader = new URLClassLoader(Seq(clientJar.toURI.toURL).toArray)
+    val sqlClassLoader: URLClassLoader = new URLClassLoader(Seq(sqlJar.toURI.toURL).toArray)
+
+    val clientClass = clientClassLoader.loadClass("org.apache.spark.sql.Dataset")
```

Comment:
HI ~ @zhenlineo @HyukjinKwon , there may be some problems with this test case, I add some logs as follows:
      https://github.com/apache/spark/compare/master...LuciferYang:spark:CompatibilitySuite?expand=1

<img width="807" alt="image" src="https://user-images.githubusercontent.com/1475305/221517891-ddb4c0f6-242a-4f48-a796-f29482dfe798.png">

From the log, both `clientClass` and `sqlClass` are loaded from `file:/home/runner/work/spark/spark/connector/connect/client/jvm/target/scala-2.12/spark-connect-client-jvm_2.12-3.5.0-SNAPSHOT.jar`, and the contents of `newMethods` and `oldMethods` are the same ...

https://pipelines.actions.githubusercontent.com/serviceHosts/c184045e-b556-4e78-b8ef-fb37b2eda9a3/_apis/pipelines/1/runs/62963/signedlogcontent/14?urlExpires=2023-02-27T08%3A53%3A13.6716136Z&urlSigningMethod=HMACV1&urlSignature=XkRKqix4ZapzEeczn7ZhWpAFhSnwWW74UX%2BKUhocftc%3D

<img width="1460" alt="image" src="https://user-images.githubusercontent.com/1475305/221518426-4e6e6274-8863-4852-b54a-2e3a471e80cf.png">

Reply:
At present, using this way to check, at least 4 apis should be incompatible:

```
private[sql] def withResult[E](f: SparkResult => E): E
def collectResult(): SparkResult
private[sql] def analyze: proto.AnalyzePlanResponse
private[sql] val plan: proto.Plan
```

Because when using Java reflection, the above 4 methods will be identified as public apis, even though three of them are `private [sql]`, and these four apis do not exist in the Dataset of the sql module:

```
public java.lang.Object org.apache.spark.sql.Dataset.withResult(scala.Function1)$
public org.apache.spark.sql.connect.client.SparkResult org.apache.spark.sql.Dataset.collectResult()$
public org.apache.spark.connect.proto.AnalyzePlanResponse org.apache.spark.sql.Dataset.analyze()$
public org.apache.spark.connect.proto.Plan org.apache.spark.sql.Dataset.plan()$
```

Reply:
also cc @hvanhovell

Reply:
Thanks so much for looking into this. The dataset test is not as important as Mima test. I will check if we can fix the issue you found. Otherwise it should be safe to delete as the test is already covered by mima.

Reply:
Thanks @zhenlineo

Reply:
Thank you for the investigation, @LuciferYang .